src/core/hle/service/hid/hidbus/ringcon.cpp

```diff
@@ -0,0 +1,278 @@
+    constexpr s16 range = 2500;
+
+    ringcon_sensor_value.data = static_cast<s16>(stick_x_f * range) + idle_value;
+
+    return ringcon_sensor_value;
+}
+
+u8 RingController::GetDeviceId() const {
+    return DEVICE_ID;
+}
+
+std::vector<u8> RingController::GetReply() {
+    const RingConCommands current_command = command;
+    command = RingConCommands::Error;
+
+    switch (current_command) {
+    case RingConCommands::GetFirmwareVersion:
+        return GetFirmwareVersionReply();
+    case RingConCommands::c20105:
+        return GetC020105Reply();
+    case RingConCommands::ReadTotalPushCount:
+        return GetReadTotalPushCountReply();
+    case RingConCommands::ReadUnkCal:
+        return GetReadUnkCalReply();
+    case RingConCommands::ReadFactoryCal:
+        return GetReadFactoryCalReply();
+    case RingConCommands::ReadUserCal:
+        return GetReadUserCalReply();
+    case RingConCommands::ReadId:
+        return GetReadIdReply();
+    default:
+        return GetErrorReply();
+    }
+}
+
+bool RingController::SetCommand(const std::vector<u8>& data) {
+    if (data.size() != 4) {
+        LOG_ERROR(Service_HID, "Command size not supported {}", data.size());
+        command = RingConCommands::Error;
+        return false;
+    }
+
+    // There must be a better way to do this
+    const u32 command_id = data[0] + (data[1] << 8) + (data[2] << 16) + (data[3] << 24);
+    constexpr std::array<RingConCommands, 7> supported_commands = {
+        RingConCommands::GetFirmwareVersion,
+        RingConCommands::ReadId,
+        RingConCommands::c20105,
+        RingConCommands::ReadUnkCal,
+        RingConCommands::ReadFactoryCal,
+        RingConCommands::ReadUserCal,
+        RingConCommands::ReadTotalPushCount,
+    };
+
+    for (RingConCommands cmd : supported_commands) {
+        if (command_id == static_cast<u32>(cmd)) {
+            command = cmd;
+            return true;
+        }
+    }
+
+    LOG_ERROR(Service_HID, "Command not implemented {}", command_id);
+    command = RingConCommands::Error;
+    return false;
+}
+
+void RingController::Enable(bool enable) {
+    ringcon_enabled = enable;
+}
+
+bool RingController::IsEnabled() const {
+    return ringcon_enabled;
+}
+
+bool RingController::IsPollingMode() const {
+    return polling_mode_enabled;
+}
+
+JoyPollingMode RingController::GetPollingMode() const {
+    return polling_mode;
+}
+
+void RingController::SetPollingMode(JoyPollingMode mode) {
+    polling_mode = mode;
+    polling_mode_enabled = true;
+}
+
+void RingController::DisablePollingMode() {
+    polling_mode_enabled = false;
+}
+
+void RingController::SetTransferMemoryPointer(u8* t_mem) {
+    is_transfer_memory_set = true;
+    transfer_memory = t_mem;
+}
+
+u8 RingController::GetCrcValue(const std::vector<u8>& data) const {
+    u8 crc = 0;
+    for (std::size_t index = 0; index < data.size(); index++) {
+        for (u8 i = 0x80; i > 0; i >>= 1) {
+            bool bit = crc & 0x80;
```

Comment:
```suggestion
            bool bit = (crc & 0x80) != 0;
```

Otherwise this truncates to a bool